src/rgw/driver/rados/rgw_data_sync.cc

```diff
@@ -2590,19 +2590,28 @@ class RGWDefaultDataSyncModule : public RGWDataSyncModule {
 
 class RGWDefaultSyncModuleInstance : public RGWSyncModuleInstance {
   RGWDefaultDataSyncModule data_handler;
+  bool start_inc{false};
+  std::string my_zonegroup;
 public:
-  RGWDefaultSyncModuleInstance() {}
+  RGWDefaultSyncModuleInstance(const RGWZoneGroup& zonegroup) {
+    start_inc = zonegroup.supports(rgw::zone_features::cross_zonegroup_sync_inc);
+    my_zonegroup = zonegroup.get_id();
+  }
   RGWDataSyncModule *get_data_handler() override {
     return &data_handler;
   }
   bool supports_user_writes() override {
     return true;
   }
+  bool should_full_sync(const RGWBucketInfo& source_info) const override {
+    // always init with full sync when the bucket is in my zonegroup (force symmetry within the zonegroup)
+    return source_info.zonegroup == my_zonegroup || !start_inc;
```

Comment:
can we find a way to make this specific to bucket replication policy instead of cross-zonegroup?

`RGWRunBucketSourcesSyncCR` may spawn several `RGWSyncBucketCR`s. some of those may correspond to cross-bucket replication from policy, but one should be the same-bucket replication that we run by default for disaster recovery. i think we need to a way to control full sync differently for those two cases

so i don't think `RGWDefaultSyncModuleInstance::should_full_sync()` is the right place for this check, unless it can take `should_full_sync(source_info, dest_info)` and compare the buckets for equality. however, i _think_ rgw supports bucket replication policy that names the same source bucket as the destination - do you know if aws allows that?

Reply:
> do you know if aws allows that?

No it doesn't, I had that check proposed here https://github.com/ceph/ceph/pull/59911/commits/888ddcbde7717ead3322077f18776dafcfed98b3.

Basically, currently, the way the pipes are considered by `RGWGetBucketPeersCR` are two types:
- a sync policy (either on the bucket or zonegroup - bear in mind even without any sync policy `RGWBucketSyncPolicyHandler()` will create a default sync policy https://github.com/ceph/ceph/blob/90a5d4efe2ef12afd016a6a1df154a5cb88db987/src/rgw/driver/rados/rgw_bucket_sync.cc#L758-L761) containing no dest bucket results in `RGWBucketSyncFlowManager::reflect()` to reflect a policy pointing to the same bucket in the current RGW's zone running sync.
- a hint created by `RGWSI_Bucket_Sync_SObj::handle_bi_update()` when a sync policy is created for a bucket pointing to that bucket.

So in all cases, we always deal with sync policies and distinguish between "bucket replication" or "within zonegroup - between zone replication) is kinda tricky now but it depends on what approach we want to take in the discussion above.

Reply:
> No it doesn't, I had that check proposed here [888ddcb](https://github.com/ceph/ceph/commit/888ddcbde7717ead3322077f18776dafcfed98b3).

okay, i think that helps. if we enforce that source != dest for new policy, then `should_full_sync(source_info, dest_info)` should be sufficient. we'll continue doing full sync for existing policy where they're equal, and the full sync of new policy will depend on the zonegroup feature. does that make sense?

Reply:
given that we deprecated the only sync module that overrode `should_full_sync()`, i'm skeptical that we want to leave this behavior up to the sync module interface. we might just hard-code that bucket comparison in/under RGWRunBucketSourcesSyncCR

Reply:
> we'll continue doing full sync for existing policy where they're equal, and the full sync of new policy will depend on the zonegroup feature. does that make sense?

Do we see a sync policy with `src_bucket == dest_bucket` but carrying filters (prefixes and/or tags) symmetry or not?

Reply:
> Do we see a sync policy with `src_bucket == dest_bucket` but carrying filters (prefixes and/or tags) symmetry or not?

if it results in different contents on each zone, then it's not symmetric. but the reason to continue doing full sync in this case is for backward compatiblity, not symmetry

Reply:
OK, let me break it down here:
1. We reject policy creation (through API and `radosgw-admin`) for a bucket that results in replicating to a bucket with the same name (here we don't care about zonegroup as cross-zonegroup logically we can't so both cases should be rejected).
2. we would change how the sync policy on the zonegroup level would work. basically, it should only control the cross-zonegroup replication not zones within the zonegroup - otherwise, that could break symmetry within the zonegroup (like having policies "enabling" replication of any source to a single bucket with a different name or stuff like that).
3. at this point we can say the sync policy that `RGWGetBucketPeersCR()` gives us is either pointing to the same bucket name with different zones in its zonegroup or a different bucket name (here doesn't matter if zonegroup is the same - although we don't support SRR yet so perhaps we need to return HTTP 501 on policy creation for that).
4. So we do if `src_bucket == dest_bucket` then it's symmetry and ignores the zonegroup feature and relies on `should_full_sync()` else consider zonegroup feature with `should_full_sync()`.

Did I miss anything?

Reply:
nice summary, i think we're on the same page!